plugins/pulp_docker/plugins/importers/sync.py

```diff
@@ -300,3 +300,51 @@ def process_main(self):
                                                       tag_name=tag, manifest_digest=manifest.digest)
             if new_tag:
                 repository.associate_single_unit(self.get_repo().repo_obj, new_tag)
+
+
+class TokenAuthDownloadStep(publish_step.DownloadStep):
+    """
+    Download remote files. For v2, this may require a bearer token to be used. This step attempts
+    to download files, and if it fails due to a 401, it will retrieve the auth token and retry the
+    download.
+    """
+
+    def __init__(self, step_type, downloads=None, repo=None, conduit=None, config=None,
+                 working_dir=None, plugin_type=None, description=''):
+        """
+        Initialize the step, setting its description.
+        """
+
+        super(TokenAuthDownloadStep, self).__init__(
+            step_type, downloads=downloads, repo=repo, conduit=conduit, config=config,
+            working_dir=working_dir, plugin_type=plugin_type)
+        self.description = _('Downloading remote files')
+        self.token = None
+
+    def process_main(self, item=None):
+        """
+        Overrides the parent method to get a new token and try again if response is a 401.
+        """
+        # Allow the original request to be retrieved from the url.
+        self._requests_map = {}
+        for request in self.downloads:
+            self._requests_map[request.url] = request
+
+        for request in self.downloads:
+            if self.token:
+                token_util.add_auth_header(request, self.token)
+            self.downloader.download_one(request, events=True)
+
+    def download_failed(self, report):
+        """
+        If the download fails due to a 401, attempt to retreive a token and try again.
+        """
+        if report.error_report.get('response_code') == 401:
+            _logger.warn(_('Download unauthorized, attempting to retrieve a token.'))
```

Comment:
Since this is a completely expected circumstance when talking to docker hub, I don't think it should be logged at the warning level. In this use case, there is nothing for the user to be worried about and no action they can take to improve the situation.